iot_config/src/gateway_info.rs

```diff
@@ -0,0 +1,176 @@
+    pub is_full_hotspot: bool,
+}
+
+impl GatewayInfo {
+    pub fn chain_metadata_to_info(
+        meta: db::IotMetadata,
+        region_map: &region_map::RegionMapReader,
+    ) -> Self {
+        let metadata = if let (Some(location), Some(elevation), Some(gain)) =
+            (meta.location, meta.elevation, meta.gain)
+        {
+            if let Ok(region) = h3index_to_region(location, region_map) {
+                Some(GatewayMetadata {
+                    location,
+                    elevation,
+                    gain,
+                    region,
+                })
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+
+        Self {
+            address: meta.address,
+            is_full_hotspot: meta.is_full_hotspot,
+            metadata,
+        }
+    }
+}
+
+fn h3index_to_region(
+    location: u64,
+    region_map: &region_map::RegionMapReader,
+) -> anyhow::Result<Region> {
+    hextree::Cell::from_raw(location)
+        .map(|cell| region_map.get_region(cell))?
+        .ok_or(anyhow!("invalid region"))
+}
+
+#[async_trait::async_trait]
+pub trait GatewayInfoResolver {
+    type Error;
+
+    async fn resolve_gateway_info(
+        &mut self,
+        address: &PublicKeyBinary,
+    ) -> Result<Option<GatewayInfo>, Self::Error>;
+
+    async fn stream_gateways_info(&mut self) -> Result<GatewayInfoStream, Self::Error>;
+}
+
+impl From<GatewayInfoProto> for GatewayInfo {
+    fn from(info: GatewayInfoProto) -> Self {
+        let metadata = if let Some(metadata) = info.metadata {
+            u64::from_str_radix(&metadata.location, 16)
+                .map(|location| GatewayMetadata {
+                    location,
+                    elevation: metadata.elevation,
+                    gain: metadata.gain,
+                    region: metadata.region(),
+                })
+                .ok()
+        } else {
+            None
+        };
+        Self {
+            address: info.address.into(),
+            is_full_hotspot: info.is_full_hotspot,
+            metadata,
+        }
+    }
+}
+
+impl TryFrom<GatewayInfo> for GatewayInfoProto {
+    type Error = hextree::Error;
+
+    fn try_from(info: GatewayInfo) -> Result<Self, Self::Error> {
+        let metadata = if let Some(metadata) = info.metadata {
+            Some(GatewayMetadataProto {
+                location: hextree::Cell::from_raw(metadata.location)?.to_string(),
+                elevation: metadata.elevation,
+                gain: metadata.gain,
+                region: metadata.region.into(),
+            })
+        } else {
+            None
+        };
+        Ok(Self {
+            address: info.address.into(),
+            is_full_hotspot: info.is_full_hotspot,
+            metadata,
+        })
+    }
+}
+
+pub(crate) mod db {
+    use futures::stream::{Stream, StreamExt};
+    use helium_crypto::PublicKeyBinary;
+    use sqlx::{PgExecutor, Row};
+
+    pub struct IotMetadata {
```

Comment:
slightly confusing having both GatewayMetadata and IotMetadata structs

Reply:
i had hoped scoping the `IotMetadata` struct inside the `db` submodule would help with that, but i couldn't think of a better way to do unless i skipped the `FromRow` impl altogether; since the `GatewayInfo` requires the hextree map to look up a Region from the asserted location and i can't change the `from_row` function signature to pass that in i needed the intermediary data structure

Reply:
I understand andymck's point, but I can't think of a better way either and I think scoping `IotMetadata` to the `db` submodule does help.